### PR TITLE
[iOS] Add an internal setting to allow the keyboard to dismiss after tapping

### PR DESCRIPTION
**Pull request by whsieh**
#### 26bb7484b4f6853b76ff5d3b2b843668f38ab019
<pre>
[iOS] Add an internal setting to allow the keyboard to dismiss after tapping
<a href="https://bugs.webkit.org/show_bug.cgi?id=277551">https://bugs.webkit.org/show_bug.cgi?id=277551</a>
<a href="https://rdar.apple.com/130552664">rdar://130552664</a>

Reviewed by Aditya Keerthi.

Add an internal setting to make it possible to dismiss the keyboard by tapping over the page,
regardless of whether the page prevents default on touch, mouse or click events. See below for more
details.

Test: fast/forms/ios/dismiss-keyboard-after-tapping.html

* LayoutTests/fast/forms/ios/dismiss-keyboard-after-tapping-expected.txt: Added.
* LayoutTests/fast/forms/ios/dismiss-keyboard-after-tapping.html: Added.

Add a test case to verify that after focusing a text field with the new flag enabled:

1.  Tapping within the focused field does not dismiss the keyboard.
2.  Tapping another text field keeps the keyboard up.
3.  Tapping an empty part of the page that prevents `touchend`, `mousedown` and `click` dismisses
    the keyboard.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _createAndConfigureDoubleTapGestureRecognizer]):
(-[WKContentView setUpInteraction]):
(-[WKContentView cleanUpInteraction]):

Add a new gesture, `_keyboardDismissalGestureRecognizer`, that&apos;s only enabled when the new setting
is enabled, and whose only purpose is to drive keyboard dismissal. This is a new gesture recognizer
because:

a.  We should recognize a tap regardless of whether default was prevented, so we can&apos;t key off of
    the synthetic click gesture.

b.  We should not recognize in the case where the user is double tapping the page to zoom out (i.e.
    single tap recognition should be delayed by double tap).

(-[WKContentView _removeDefaultGestureRecognizers]):
(-[WKContentView _addDefaultGestureRecognizers]):
(-[WKContentView gestureRecognizer:shouldRecognizeSimultaneouslyWithGestureRecognizer:]):
(-[WKContentView gestureRecognizerShouldBegin:]):

Bail early from the keyboard dismissal gesture if any of the following are true:
1. The keyboard is not shown.
2. The tap is inside of the focused element.
3. A hidden contenteditable element has focus.
4. The tap gesture is used to stop momentum scrolling.

(-[WKContentView _keyboardDismissalGestureRecognized:]):

Asynchronously check if a tap at this location should result in keyboard dismissal using
`shouldDismissKeyboardAfterTapAtPoint` (see below).

(-[WKContentView deferringGestureRecognizer:shouldDeferOtherGestureRecognizer:]):
* Source/WebKit/UIProcess/ios/WKScrollViewTrackingTapGestureRecognizer.h: Copied from Source/WebKit/UIProcess/ios/WKSyntheticTapGestureRecognizer.h.
* Source/WebKit/UIProcess/ios/WKScrollViewTrackingTapGestureRecognizer.mm: Copied from Source/WebKit/UIProcess/ios/WKSyntheticTapGestureRecognizer.h.
(-[WKScrollViewTrackingTapGestureRecognizer reset]):
(-[WKScrollViewTrackingTapGestureRecognizer touchesBegan:withEvent:]):

Split out logic to keep track of the last touched scroll view out into a separate class, which we
use for the new gesture recognizer. This allows us to check whether the keyboard dismissal tap
gesture was recognized while momentum scrolling, in which case we don&apos;t allow it to recognize.

* Source/WebKit/UIProcess/ios/WKSyntheticTapGestureRecognizer.h:
* Source/WebKit/UIProcess/ios/WKSyntheticTapGestureRecognizer.mm:
(-[WKSyntheticTapGestureRecognizer reset]):
(-[WKSyntheticTapGestureRecognizer touchesBegan:withEvent:]): Deleted.
(-[WKSyntheticTapGestureRecognizer lastTouchedScrollView]): Deleted.
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::shouldDismissKeyboardAfterTapAtPoint):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::shouldDismissKeyboardAfterTapAtPoint):

Only dismiss the keyboard if the tap is either:
- Over a very large element responding to click events, or
- Not over an element responding to click events.

This makes it likely that the user will be able to dismiss the keyboard by tapping over blank areas
of the page, even if the page installs click handlers over a large container element, while making
it unlikely that the keyboard would unintentionally dismiss when tapping (for instance) a small,
text editing control to bold or italicize text.

Canonical link: <a href="https://commits.webkit.org/281844@main">https://commits.webkit.org/281844@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/e318cf22f657609921258a11a26f114e12937d19

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/61194 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40555 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/13775 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/65144 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11743 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/48232 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/12018 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49471 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8173 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/63228 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37726 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53021 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30301 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34406 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/10246 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/10656 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/54298 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56223 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/10541 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/66875 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/60442 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/5141 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/10351 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/56843 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/5163 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/52984 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57040 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13639 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4256 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/82197 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/36359 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/14343 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/37442 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/38536 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/37186 "Built successfully") | | |

### CODE REVIEW

**Comment by whsieh** [comment]
Hm... the test doesn't crash locally, testing against iOS 18. Maybe something's broken wrt. the edit menu and data detectors on iOS 17

**Comment by whsieh** [comment]
> ...so this is a bincompat issue due to removing `_WKWebExtension` from the WebKit framework. DDUI links SafariSharedUI, which still expects this old symbol stub because it was built against iOS 17's WebKit stack, but doesn't find it because we're linking our built WebKit :/

This binary incompatibility was only introduced yesterday, in https://commits.webkit.org/281723@main...

**Comment by whsieh** [comment]
A quick scan of EWS shows that these crashes are happening on other EWS runs as well, e.g.:

- https://ews-build.webkit.org/#/builders/47/builds/37493
- https://s3-us-west-2.amazonaws.com/ews-archives.webkit.org/ios-simulator-17-arm64-release-layout-tests/33a3be79.txt

File: Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm [hunk 1]
```diff
@@ -676,6 +677,15 @@ static WKContentMode contentMode(const TestOptions& options)
...
     [webpagePreferences setPreferredContentMode:contentMode(options)];
+
+    if (options.keyboardDismissalGestureEnabled()) {
+        for (_WKFeature *feature in WKPreferences._features) {
+            if ([feature.key isEqualToString:@"KeyboardDismissalGestureEnabled"]) {
+                [[configuration preferences] _setEnabled:YES forFeature:feature];
+                break;
+            }
+        }
+    }
```
Comment (pxlcoder) [inline Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm]: Why is this necessary? I don't see this done for other flags.

I think you might need to refer to the flag as `KeyboardDismissalGestureEnabled` (uppercase), rather than `keyboardDismissalGestureEnabled`, in `TestOptions.cpp` and the test file.

**Review by pxlcoder — APPROVED** [review]

File: Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm (hunk 1 above)
Comment (whsieh) [reply Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm]: Oh, good catch — using an uppercase `K` there ought to Just Work™ without this extra logic.

Thanks for the review!
